### ISSUE TITLE
sklearn preprocessing notes

### ISSUE THREAD

**Issue by redsnow1992** [redsnow1992/notes#35]
# Standardization和几种scaler
因为`scaler`实现了`Transformer`的ＡＰＩ，`scaler`会保存状态，可以使用`fit`和`transform`:

| scaler类型 | 变换 | 目标 |
| --- | --- | --- | 
| StandardScaler | `(X-mean(X))/std(X)` | `mean=0, std=1` |
| MinMaxScaler | `((X-X.min)/(X.max-X.min))*(max-min)+min` | `min <= x <= max` |
| MaxAbsScaler | `X/abs(max(X))` ||

特别地:
1. 对于稀疏数据,可以使用`MaxAbsScaler`或者`StandardScaler(with_mean=False)`
2. 对于带有离群点的数据,可以使用`RobustScaler`

# Normalization
将一组数据(向量)的模标准化为单位1
```python
 X = [[ 1., -1.,  2.],
      [ 2.,  0.,  0.],
      [ 0.,  1., -1.]]
preprocessing.normalize(X, norm='l2')
array([[ 0.40..., -0.40...,  0.81...],
       [ 1.  ...,  0.  ...,  0.  ...],
       [ 0.  ...,  0.70..., -0.70...]])
```

# Binarization(二值化)
根据阈值将数据分成0,1
```python
X = [[ 1., -1.,  2.],
      [ 2.,  0.,  0.],
      [ 0.,  1., -1.]]
preprocessing.Binarizer(threshold=1.1).transform(X)
array([[ 0.,  0.,  1.],
       [ 1.,  0.,  0.],
       [ 0.,  0.,  0.]])
```
# 对类别特征进行编码
1. 将类别数据进行编码
2. 1中的编码进行`OneHotEncoder`

**默认一列数据就是一个feature**
编码前的类别:

| col1 | 编码后col1 | col2 | 编码后col2 | col3 | 编码后col3 |
| --- | --- | --- | --- | --- | --- |
| `male` | 0 | `from Europe` | 0 | `uses Firefox` | 0 |
| `male` | 0 | `from US` | 1 | `uses Chrome` | 1 |
| `female` | 1 |  `from Asia` | 2|  `uses Safari` | 2 | 
| `female` | 1 | `from Asia` | 2 | `uses Internet Explorer` | 3 |

ps: `axis`参数为0表示按列来处理,1表示按行来处理    
ref: http://scikit-learn.org/stable/modules/preprocessing.html
